controllers/precache.go

```diff
@@ -0,0 +1,768 @@
+package controllers
```

Comment:
This is a new large file - not shown by github by default, hence adding this comment